design/baremetal-operator/bulk-set-bios-config.md

```diff
@@ -0,0 +1,414 @@
+the same schema, its not necessary to cache it each time for similar hardware.
+When BIOS data is read from the Ironic API for each host the following
+ actions will be taken:
+
+- If the `HostFirmwareSettings` already has a reference to a  `FirmwareSchema`
+  resource, then the schema data in the BIOS API will be ignored
+- If there is no reference, the schema data for all BIOS settings for the
+  host will be stored locally
+- A hash will be generated from this stored data, if a corresponding hash
+  does not exist in the `FirmwareSchema` CRs, a new CR will be created and
+  the schema data will be copied to it
+- The reference to the `FirmwareSchema` resource in `HostFirmwareSettings`
+  will be updated with the hash. A suggested reference string is to
+  include the vendor and model in the string, such as
+  "vendor,model,64 chars of hash"
+- Using the hash will ensure that the schemas match. In the case of
+  hardware from the same vendor and model, but with a different firmware
+  version, it is possible that the schemas would differ. In this case a new
+  `FirmwareSchema` will be created.
+- The name of the `FirmwareSchema` resource will be added to the Labels
+  for the `HostFirmwareSettings` so that it can be used to delete the
+  `FirmwareSchema` resource when it is no longer being used.
+
+An example of the proposed CRD for `HostFirmwareSettings` is as follows:
+
+```yaml
+---
+kind: CustomResourceDefinition
+metadata:
+  name: hostfirmwaresettings.metal3.io
+spec:
+  group: metal3.io
+  names:
+    kind: HostFirmwareSettings
+    listKind: HostFirmwareSettingsList
+  scope: Namespaced
+  versions:
+  - name: v1alpha1
+    schema:
+          spec:
+            description: HostFirmwareSettingsSpec defines the desired state
+            properties:
+              vendorSettings:
+                items:
+                  description: describes a setting whose value can be changed
+                  properties:
+                    name:
+                      description: Identifier for the Firmware setting.
+                      type: string
+                    value:
+                      description: Value of the Firmware setting.
+                      type: string
+                  required:
+                  - name
+                  - value
+                  type: object
+                type: array
+            type: object
+          status:
+            description: defines the observed state of HostFirmwareSettings
+            properties:
+              settingsData:
+                items:
+                  description: describes current settings and their schema
+                  properties:
+                    name:
+                      description: Identifier for the Firmware setting.
+                      type: string
+                    value:
+                      description: Value of the Firmware setting.
+                      type: string
+                    schemaReference:
```

Comment:
Wouldn't be enough to specify the schema reference per host (instead that per item)? Not sure if also would be required to specify the namespace (thinking to the case were the BMH could be spread across different namespaces, but I guess the schema will continue to stay in just one)

Reply:
That's an interesting point about the namespace, actually...

If we don't specify the namespace, it's implied that we're looking in the same namespace as the Host/Settings. In principle, that makes sense to me - it's not really safe to have Hosts in different namespaces referencing the same Schema CR for RBAC reasons.

_However_ in the use case we're thinking about (and trying to save memory in), there are hundreds of Hosts, each in its own namespace. If we always reference the schema in the local namespace, then we'll end up back with 1 per Host and we won't save any memory at all by splitting the CRDs (in fact, it'll be slightly worse).

The "correct" solution to this is [API decomposition](https://docs.google.com/document/d/1FL8oA0_WNcPdiC-0zrToinl8aol-NPQUjWe-OkcvTqA/edit#heading=h.lmpo60bkt5pm) but this is not on the cards any time soon.

The best hack I can think of right now would be to have an optional CLI flag for the controller to use a particular namespace for the Schema instead of the local one. So I think do we need to store a namespace in the CR after all.

Reply:
Or assuming that anyhow BMO is the only actor responsible to create/update (and maybe delete) the Firmware Schema resources, then they could be kept in the current [POD_NAMESPACE](https://github.com/metal3-io/baremetal-operator/blob/399f5ef7ee3831014c1425250bc4fa49641a8709/main.go#L106) configured - this could simplifies a little bit the RBAC settings and wouldn't require any additional namespace field in the Schema CR, as I originally thought

Reply:
Hardware admins need to be able to read it though (we're creating it for their benefit, not so much BMO's), and I don't think we should assume that in general they'll have RBAC rights to the same namespace that the BMO pod is running in.
We also need to be pretty clear about telling them where to look for it.

Reply:
OK, I added a `schemaNamespace` hint to the HFS CRD for this case. I also changed the FirmwareSchema CRD to be scope `Namespaced`.

Reply:
Ok, it seems fine to keep the namespace reference. Do we already have a definition for the "hardware admin" role somewhere? Or do we need to clarify it better here?

Reply:
It might be helpful if we wrote up a doc with a bunch of personas that we expect to interact with metal³ (it would certainly make writing user stories easier) but I don't think we have one yet.

In any moderately sized organisation I'd expect there to be different people responsible for maintaining the hardware fleet vs. deploying apps on k8s. In a multitenant situation with multiple spoke clusters being provisioned from a central hub, those are _definitely_ different people.

Reply:
Even though not strictly related, while looking for some materials on the multi-tenancy I stumbled upon https://github.com/kubernetes-sigs/multi-tenancy/blob/master/docs/profiles/profile-soft-multitenancy-s1.md, which has an interesting personas definition